### PR TITLE
UDP Ipv6 support

### PR DESCRIPTION
**Pull request by simias**
Far from complete but I could use some feedback

### CODE REVIEW

File: src/udp_address.cpp [hunk 1]
```diff
@@ -129,88 +123,55 @@ int zmq::udp_address_t::resolve (const char *name_, bool bind_)
...
-unsigned short zmq::udp_address_t::family () const
-#else
-sa_family_t zmq::udp_address_t::family () const
-#endif
+int zmq::udp_address_t::to_string (std::string &addr_)
 {
-    return AF_INET;
+    // XXX what do (factor TCP code?)
+    addr_ = address;
+    return 0;
```
Comment (simias) [inline src/udp_address.cpp]: What exactly are the requirements for this `to_string` method? The current implementation just returns the string passed to the resolver but the TCP implementation actually bothers re-generating the string from the sockaddr values.

File: src/udp_engine.cpp [hunk 2]
```diff
@@ -111,9 +111,11 @@ void zmq::udp_engine_t::plug (io_thread_t *io_thread_, session_base_t *session_)
...
         if (!options.raw_socket) {
-            out_address = address->resolved.udp_addr->dest_addr ();
-            out_addrlen = address->resolved.udp_addr->dest_addrlen ();
+            const ip_addr_t *out = address->resolved.udp_addr->target_addr ();
+            out_address = out->as_sockaddr ();
+            out_addrlen = out->sockaddr_len ();
         } else {
+            /// XXX fixme ?
             out_address = (sockaddr *) &raw_address;
             out_addrlen = sizeof (sockaddr_in);
```
Comment (simias) [inline src/udp_engine.cpp]: I don't quite understand how this `raw_address` part fits in, I see that the address is parsed within this very file in `out_event`, where does the string come from exactly? Why doesn't it use the `udp_address` parser?

File: src/udp_engine.cpp [hunk 3]
```diff
@@ -145,11 +149,29 @@ void zmq::udp_engine_t::plug (io_thread_t *io_thread_, session_base_t *session_)
...
+                mreq.imr_multiaddr = mcast_addr->ipv4.sin_addr;
+                mreq.imr_interface = bind_addr->ipv4.sin_addr;
+
+                rc = setsockopt (fd, IPPROTO_IP, IP_ADD_MEMBERSHIP,
+                                 (char *) &mreq, sizeof (mreq));
+            } else if (mcast_addr->family () == AF_INET6) {
+                struct ipv6_mreq mreq;
+
+                mreq.ipv6mr_multiaddr = mcast_addr->ipv6.sin6_addr;
+                mreq.ipv6mr_interface = 0; // XXX fixme?
```
Comment (simias) [inline src/udp_engine.cpp]: So here I'm having a bit of a conundrum: unlike the IPv4 code `struct ipv6_mreq` doesn't let you specify the IP address of the interface to bind but rather the interface number. It's trivial to lookup if we have the interface name using `if_nametoindex` but of course at this point we don't have the interface name anymore and on top of that ZMQ URLs lets you specify an interface by IP: `udp://[abcd::1];[ff00::1]:5555`.

The easy solution would be to only support specifying a source interface by giving the interface name but that's obviously limiting and not very consistent. The hard solution would be to have a method to look up an interface index based on an IP (pretty much the opposite of `zmq::ip_resolver_t::resolve_nic_name`). Of course the tedious part is that this code is very platform-specific.

I think the latter solution is the way to go but I'd like to have some feedback on that.

File: unittests/unittest_resolver_common.hpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#ifndef __UNITTEST_RESOLVER_COMMON_INCLUDED__
```
Comment (bluca) [inline unittests/unittest_resolver_common.hpp]: a copyright header is needed here

File: src/udp_engine.cpp (hunk 3 above)
Comment (bluca) [reply src/udp_engine.cpp]: If you want to start by only supporting the interface name, and then later on add support for the IP, that's fine given this socket type is a work in progress

File: src/udp_engine.cpp (hunk 2 above)
Comment (bluca) [reply src/udp_engine.cpp]: not sure about the address, but thw raw_socket option is used to read/write UDP datagrams without any ZMTP headers

**Comment by simias** [comment]
I cleaned up the code, added tests. IPv6 multicast works but only if you bind `*` or an interface name, binding by interface IP returns `ENODEV`. I haven't touched the raw address code so it still only supports IPv4, I suspect that it wouldn't be too hard to modify to use `ip_resolver` instead but it would probably be a good idea to add some tests first.

I want to add multicast tests but I'd need to add a socket option to allow looping the multicast packets on the same host using `setsockopt (IP[V6]_MULTICAST_LOOP)`. I think this ought to be added as a `zmq_setsockopt` parameter but I'll leave that for a future PR. Similarly it might be worth adding support for the `IP_MULTICAST_TTL` option.

**Comment by simias** [comment]
I don't understand why some travis builds don't find my header in `unittests`, it works fine on my machine. I tried replacing `#include "unittest_resolver_common.hpp"` with `#include "../unittests/unittest_resolver_common.hpp"` to match the `testutil.hpp` include but it's still broken.

**Comment by bluca** [comment]
Doesn't the header have to be added to the executable via CMakeLists.txt?

**Comment by bluca** [comment]
The CI is already slow enough, I don't think we should remove the parallel builds - it's easy enough to do in a temporary commit when debugging is needed

**Comment by bluca** [comment]
Like this:

https://github.com/zeromq/libzmq/blob/master/tests/CMakeLists.txt#L185

**Comment by simias** [comment]
But why does it work when I build locally using cmake?

Also the unnittests CMakeLists.txt doesn't mention any include file, including `testutil.hpp`, should it?

https://github.com/zeromq/libzmq/blob/7aba6821ac56a15706896618ef35d4c9df5032a1/unittests/CMakeLists.txt#L29

File: src/udp_address.cpp (hunk 1 above)
Comment (bluca) [reply src/udp_address.cpp]: I have a feeling there was something related to disconnect/unbind - check the git log, there might be some clues

**Comment by simias** [comment]
Uh, the plot thickens:

```
CMake Error at unittests/CMakeLists.txt:29 (add_executable):
  Cannot find source file:
    unittest_resolver_common.hpp
  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
```

Does it mean that the buildbot hasn't checked-out that file for some reason?

**Comment by bluca** [comment]
gah - it's because the CI tests uses the distributable tarball. You need to add the new header to automake, then it will work

**Comment by simias** [comment]
Looks better. If you don't have any objections maybe that's enough changes for this PR and the rest of the work (IPv6 raw sockets, to_string, LOOP support etc...) can be left for later.

**Comment by simias** [comment]
Thanks

File: src/udp_address.cpp [hunk 5]
```diff
@@ -79,24 +77,38 @@ int zmq::udp_address_t::resolve (const char *name_, bool bind_)
...
 
+        //  This is a hack because we need the interface index when binding
+        //  multicast IPv6, we can't do it by address. Unfortunately for the
+        //  time being we don't have a generic platform-independent function to
+        //  resolve an interface index from an address, so we only support it
+        //  when an actual interface name is provided.
+        if (src_name == "*") {
+            bind_interface = 0;
+        } else {
+            bind_interface = if_nametoindex (src_name.c_str ());
```
Comment (AntonioGMuriana) [inline src/udp_address.cpp]: This breaks again compilation for Universal Windows App.
Reopen issues #2309

Comment (AntonioGMuriana) [inline src/udp_address.cpp]: This breaks again compilation for Universal Windows App.
Reopen issue #2309

**Pull request merged**